SRPG-29 Feat: CharacterGrowth 분리 작업

jhwoo944 opened the pull request:
## 개요
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
CharacterGrowth에서 CharacterGrowthManager를 분리
 - CharacterGrowth : 캐릭터의 개인 고유 성장 데이터를 저장함
 - CharacterGrowthManager : 캐릭터의 각종 스탯 및 필드를 계산 및 처리하는 메서드를 저장함 
 - CharacterGrowthManager는 Managers에 등록되어 있으며, 모든 캐릭터가 공용으로 사용하고 Init으로 처리할 character를 받아옵니다.

 - 모든 SO(캐릭터/스킬/특성/클래스/아이템)의 ID 값 자료형을 int로 수정했습니다.
 - PlayerData의 UID값은 string인데 int로 변경할지 논의 필요.
 - 그 외 피드백 받은 부분 수정
<!---- Resolves: #(Isuue Number) -->

## PR 유형
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [x] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [x] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다. (ex. feature/SRPG-11-UIBase)
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

jhwoo944 committed:
SRPG-29 Feat: CharacterGrowth 분리 작업

CharacterGrowth에서 CharacterGrowthManager를 분리
 - CharacterGrowth : 캐릭터의 개인 고유 성장 데이터를 저장함
 - CharacterGrowthManager : 캐릭터의 각종 스탯 및 필드를 계산 및 처리하는 메서드를 저장함
CharacterGrowthManager는 Managers에 등록되어 있으며, 모든 캐릭터가 공용으로 사용하고 Init으로 처리할 character를 받아옵니다.

 - 모든 SO(캐릭터/스킬/특성/클래스/아이템)의 ID 값 자료형을 int로 수정했습니다.
 - PlayerData의 UID값은 string인데 int로 변경할지 논의 필요.
 - 그 외 피드백 받은 부분 수정

GeCrabKani reviewed (changes requested):
CharacterGrowth의 maxLevel get 부분에서  if (Limit > 1)로 되어있습니다.
Limit = 1일 때 조건을 통과하지 못합니다.

CharacterGrowth의 Exp set에서 level을 사용하고 있습니다.
이 경우 Level의 set이 불려지지 않아 레벨업을 할 수 없습니다.
level -> Level로 변경 필요

jhwoo944 commented:
> CharacterGrowth의 maxLevel get 부분에서 if (Limit > 1)로 되어있습니다. Limit = 1일 때 조건을 통과하지 못합니다.
> 
> CharacterGrowth의 Exp set에서 level을 사용하고 있습니다. 이 경우 Level의 set이 불려지지 않아 레벨업을 할 수 없습니다. level -> Level로 변경 필요

확인하고 수정하겠습니다
＊ 수정했습니다.

jhwoo944 committed:
SRPG-29 Fix: 버그 수정

 - CharacterGrowth.cs
1.
 - maxLevel get 부분 : if (Limit > 1) →  if (Limit >= 1)
2.
 - Exp set
 - level → Level

jhwoo944 commented:
이제보니 CharacterGrowth 브랜치에 해야할 작업을 실수로 AccountData에 했네요
주의하겠습니다;

GeCrabKani reviewed (approved):
이렇게 할 경우 캐릭터의 스탯에 변화를 주려할 때마다 CharacterGrowthManager Init을 해줘야겠네요
사용에 불편함은 조금 있겠지만 성능 상 문제는 크지 않을 것 같긴 합니다.

jhwoo944 commented:
> 이렇게 할 경우 캐릭터의 스탯에 변화를 주려할 때마다 CharacterGrowthManager Init을 해줘야겠네요 사용에 불편함은 조금 있겠지만 성능 상 문제는 크지 않을 것 같긴 합니다.

캐릭터 개체 별로 성장 데이터는 각자 필수로 갖고 있어야 하지만 메서드 클래스는 각자 갖고 있을 필요없이 공용으로 하나만 존재하는 편이 더 나을 것 같아서 이렇게 고쳐봤는데, 각자 갖고 있는 게 더 낫다면 그렇게 수정할게요

자유롭게 의견 남겨 주세요

GeCrabKani commented:
보니까 클래스 내의 함수들은 코드 세그먼트에 저장이 되고 모든 객체들이 해당 함수를 공용으로 사용한다고 하네요
따라서 클래스 내부에 존재해도 공용의 함수 하나로 실행된다고 합니다.

Character.cs가 ChracterGrowth.cs를 들고 있고, CharacterGrowth.cs가 Character.cs를 들고 있는 구조는 순환 참조로 설계상 좋지 않습니다.

해당 구조는 저번에 튜터님이 얘기해주신대로 DB 작업이 끝난 후 구조를 수정하는 것이 좋아보입니다.

jhwoo944 commented:
> 보니까 클래스 내의 함수들은 코드 세그먼트에 저장이 되고 모든 객체들이 해당 함수를 공용으로 사용한다고 하네요 따라서 클래스 내부에 존재해도 공용의 함수 하나로 실행된다고 합니다.
> 
> Character.cs가 ChracterGrowth.cs를 들고 있고, CharacterGrowth.cs가 Character.cs를 들고 있는 구조는 순환 참조로 설계상 좋지 않습니다.
> 
> 해당 구조는 저번에 튜터님이 얘기해주신대로 DB 작업이 끝난 후 구조를 수정하는 것이 좋아보입니다.

그럼 CharacterGrowthManager는 Managers에서 등록 해제하고 Character가 들고 있도록 수정하겠습니다.
ChracterGrowth 순환참조 구조는 그럼 말씀하신대로 DB 작업 끝나는 대로 수정 할게요

SeRDic0705 commented on Assets/Scripts/Objects/Characters/CharacterGrowthManager.cs:
현우님 커밋에서 항상 주석이 깨져보이는데, 비주얼스튜디오 에디터의 기본 인코딩설정이 UTF-8로 설정되어있는지 다시 한번 확인 부탁드려요
